Py bindings ci

0xBEEFCAF3 opened the pull request:
closes #3
closes #57

DanGould commented:
We should follow [bdk-ffi's test-python GitHub workflow](https://github.com/bitcoindevkit/bdk-ffi/blob/master/.github/workflows/test-python.yaml) as an example

DanGould commented:
I see BDK's generate scripts, at least [the linux one for python](https://github.com/bitcoindevkit/bdk-ffi/blob/master/bdk-python/scripts/generate-linux.sh) do build the lib before running `uniffi-bindgen`. That's probably why CI is still failing.

0xBEEFCAF3 commented:
Finally done wrestling CI. Major changes include:
* Removing support for window bindings
* Skipping broken Python tests (all execept one) and the int. tests. Running the tests are still useful b/c they show there are no build time errors.
* Missing features flags when building FAT Macos binary

cc @DanGould @spacebear21

DanGould commented:
The one thing I would hope for here is that the CI changes are squashed together so that intermediate commits won't fail the CI checks they implement

DanGould commented:
yes

0xBEEFCAF3 commented:
> yes

Moved the commit which adds the CI jobs as the last commit and squashed relevant commits. 

The other commits atomically fix different things, not neccecarily for CI. For example removing windows support and fixing the x86 MacOS binary. Even without the last commit we would want these and they achieve independently useful things. I could make a monster commit that fixes the various problems that exist in main and then a CI commit but I do think there is an added benefit to having seperate atomic commits. For instance we could revert just the commit that removes window support in the future. Let me know how you would proceed

DanGould commented on python/scripts/bindgen_generate.sh:
```diff
@@ -2,11 +2,9 @@
 
 #!/bin/bash
 chmod +x ./scripts/generate_linux.sh
-chmod +x ./scripts/generate_windows.sh
```
Do you wish not also to remove the generate_windows script?

0xBEEFCAF3 replied:
Agh Yeah this was done in a seperate commit. I will move that change to this commit

0xBEEFCAF3 replied:
Done

DanGould commented on .github/workflows/build-python.yml:
```diff
@@ -0,0 +1,100 @@
+      - name: "Generate payjoin-ffi.py and binaries"
+        run: bash ./scripts/generate_macos.sh
+
+      - name: "Build wheel"
+        # Specifying the plat-name argument is necessary to build a wheel with the correct name,
+        # see issue BDK#350 for more information
+        run: python3 setup.py bdist_wheel --plat-name macosx_11_0_x86_64 --verbose
```
Should we mention, as bdk does, the rationale for testing only the x86 wheel? https://github.com/bitcoindevkit/bdk-ffi/blob/master/.github/workflows/test-python.yaml#L97

0xBEEFCAF3 replied:
good comment. I actually realized that we are not building the arm64 wheel on CI. Will add this step.

0xBEEFCAF3 replied:
Done

DanGould commented on python/scripts/generate_windows.sh:
```diff
@@ -6,8 +6,7 @@ LIBNAME=payjoin_ffi.dll
 cd ../
-cargo run --features uniffi --bin uniffi-bindgen generate src/payjoin_ffi.udl --language python --out-dir python/src/payjoin/
-
+cargo run --features uniffi --bin uniffi-bindgen generate --library target/release/$LIBNAME --language python --out-dir python/src/payjoin/
```
It would seem this windows script should have been removed with its support a few commits earlier.

0xBEEFCAF3 replied:
yes you are right

0xBEEFCAF3 replied:
Done

DanGould reviewed (commented):
Removing the windows script entirely seems not to be done.

Otherwise this is looking pretty darn good

spacebear21 reviewed (approved):
tACK 57aabafe2d70d70f69905a2a9175513d2a87c836

Tested on macOS, it took a bit of fiddling with python to get the incantations right but got build and test passing with:

```bash
python3 -m venv pjvenv
source pjvenv/bin/activate
bash scripts/generate_macos.sh
python3 setup.py bdist_wheel --verbose
pip3 install ./dist/*.whl
python3 -m unittest --verbose test/payjoin_unit_test.py

----------------------------------------------------------------------
Ran 6 tests in 0.000s

OK (skipped=5)
```

0xBEEFCAF3 commented:
> tACK [57aabaf](https://github.com/LtbLightning/payjoin-ffi/commit/57aabafe2d70d70f69905a2a9175513d2a87c836)
> 
> Tested on macOS, it took a bit of fiddling with python to get the incantations right but got build and test passing with:
> 
> ```shell
> python3 -m venv pjvenv
> source pjvenv/bin/activate
> bash scripts/generate_macos.sh
> python3 setup.py bdist_wheel --verbose
> pip3 install ./dist/*.whl
> python3 -m unittest --verbose test/payjoin_unit_test.py
> 
> ----------------------------------------------------------------------
> Ran 6 tests in 0.000s
> 
> OK (skipped=5)
> ```

@spacebear21 thanks for testing! Was this on an arm or x86 mac?

spacebear21 commented:
Ah yes I should've specified, this was on ARM.

DanGould reviewed (approved):
ack 76282f1b63de77c2315cb1140af396a8621e293d
